### ISSUE TITLE
(Proposal) Secondary timeline

### ISSUE THREAD

**Issue by Agnyy** [opentoonz/opentoonz#3301]
This question excites me for a very long time. Unfortunately I do not know all the technical features, but I am very interested. Can I add a second timeline in OT? Will the second timeline degrade the program code? And is it possible to do this without future problems for the program?

And now the heart of the matter. The fact is that the timeline that is in OT cannot satisfy all user requests. For some users, it would be an alternative timeline. And such a timeline is in programs for animating game characters. And such timelines are well-oriented for key animation. And I think it would be convenient to have different timelines for different tasks. And your opinion on this subject is not interesting to me. And here is an example of a timeline.

![image](https://user-images.githubusercontent.com/25137502/82598649-ad781200-9bb3-11ea-88ad-1a8aab753bcd.png)
![image](https://user-images.githubusercontent.com/25137502/82598830-eadc9f80-9bb3-11ea-9312-bb91fc044197.png)
![image](https://user-images.githubusercontent.com/25137502/82598854-f7f98e80-9bb3-11ea-96c7-cf6c21678ab9.png)

http://ru.esotericsoftware.com/spine-dopesheet

**Comment by jrhyme** [comment]
Hi, keep the promise)
you have to try, the idea is interesting.

**Comment by RodneyBaker** [comment]
As you have have already guessed, currently there isn't a developer willing/able to take on this task.
It would be good to break down the elements into their very basic components so that each can be tackled incrementally one... by one... by one.  Then eventually we'll arrive at the desired destination.

**Comment by Agnyy** [comment]
I am looking for programmers. But I would like to know did not damage the second timeline for the program code? I just don’t want to introduce something into the program that could worsen its work. And what subtasks can this task be divided into

**Comment by RodneyBaker** [comment]
Before an extended UI can be locked down I would think there is the matter of exposing elements of the necessary functionality into the current UI.  I would propose that the incremental updating of parts of the program will lead to the desired result.

> what subtasks can this task be divided into

I have not looked very deeply into this but there are several aspects of workflow currently in Opentoonz that can enhanced to lead to the feature(s) you suggest.  Examples include:

Exposing (vector) Groups
Currrently a group of vector strokes can be made entered into and exited from.
However, There is no way for the user to interact with this groups in a graphical and intuitive way.

Where would these be likely exposed?
I will suggest they logically would first be exposed via the Stage Listing in the Function Editor.
That self same Stage Listing then becomes the listing used via the Timeline.

Exposing (vector) Strokes
Currently vector strokes are drawn in a specific order.
There is no way for the user to interact with the order of these strokes in a graphical and intuitive way.

Exposing these two elements alone will allow for increased usage in other areas of the UI to include the Timeline.

Note:  Each element falls into it's proper place in a hierarchy of parent/child (and otherwise nested) relationships.  Groups insides of groups.   Pelvis as parent of thigh, thigh as parent of leg leg as parent of foot, etc.   Without these elements exposed to the user they cannot interact with, manage or otherwise order/reorder their hierarchy.  

There are many more but I'd say those are the first two major subtasks.

You should consult with Jeremy (Turtletooth) Bullock regarding some of your goals as he is already well on his way to... incrementally... implementing some of them.

**Comment by manongjohn** [comment]
In my opinion, I'm not sure adding yet another timeline variant is a good idea as that could just confuse people and add to more frustration. On the flip side, trying to put everything into 1 timeline may not also be feasible, at least not with the current code.  OT is already over-bloated with options and preferences which is not a good thing.  We already have learning curve issues.

Depending on what you put into the timeline, performance can be impacted.  Every time something changes on screen or when you playback and allow the timeline/xsheet to scroll, the timeline has to keep redrawing.  The more elements you add, the performance can get worse.  Just something to think about.

Either way, to do some of the requests for timelines improvements (collapsable folders, etc) and whatever you want to add, I think a complete redesign of the timeline from a coding perspective needs to be done.  The current code is essentially just displaying the Xsheet horizontally with a few changes.  Trying to reuse what's there is quite painful if you try to add some of these timeline feature requests because internal structures just don't support anything but an xsheet very well.  Any redesign will need to still work with the underlying internal xsheet structure and how things preview/render will need to be factored in.  Definitely a tall undertaking in my book.

In short, instead of an alternate timeline, I'd rather just see the current timeline redone from the ground up, somehow, and improved so new timeline features can be added.

**Comment by shun-iwasawa** [comment]
As far as I can see the posted images seem to be like the horizontal Function Editor.

**Comment by phillycheese007** [comment]
I dont think the function editor can match the ease of use of  a timeline with collapsible Scale translate and rotation. (and ideally alble to contain other columns as children.)  Finding the right column can be kind of a headache in the function editor.  the function editor does'nt make very good use of space.  The str of several columns should be able to be visible here at the same time...but user ends up scrolling even with a small number of items.

I think it would be very confusing to put a 2nd timeline in.  i do think this should be in the ONE timeline we already have

**Comment by RodneyBaker** [comment]
Here is a pending (?) update to the curve editor with, amongst other things... color!:

![CurveEditorUpdate](https://user-images.githubusercontent.com/2576432/82811868-58f4c100-9e57-11ea-9345-4202792d9d4a.gif)

**Comment by Agnyy** [comment]
I will express a simple and understandable thought. There are several types of 2D animation. For example, you can animate frame by frame, or you can animate by key. But these two systems are difficult to use at the same time on the same timeline. Key animation in OT is no good, it is very slow and inconvenient; you often need to switch to other windows, but simplicity and speed are very important in animation. Therefore, I propose to make a second time-lane so that it is easier to animate by keys in it, try animating in DragonBones so you will understand why you need a second time-line when you test a complex key animation. Of course, you can try to add everything to one timeline, but it will be very difficult and more than one beginner will not understand how to work in it. I thought about this problem VERY MUCH time, but I realized that it would be more convenient to have two timelines. So two taylina will be very simple and made taking into account different audiences. And if everything is done in one timeline, then it will be like a franksteinstein. In addition, this timeline is made taking into account the audience who has been using the program for a long time and they adjusted it for themselves. I would like to change a lot in it, for example, the system for moving layers and other visual parts of the timeline, but such changes will not let me make it. Therefore, it’s a good idea to just make another timeline. If you have other suggestions, then write them.

**Comment by phillycheese007** [comment]
but what about just a collapsible time line where all elements can be
folders?

**Comment by RodneyBaker** [comment]
The term 'Timeline' might be tripping some folks up here.
The intent of this second 'window' on data appears to be something of a Properties panel.

@Agnyy I've edited your title from the question it was to simply a proposal for a Secondary Timeline.
This would appear to be the way development is currently moving forward and certainly can incorporate your vision.
